### ISSUE TITLE
Performance: avoid locking in Pool?

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#147]
Noticed that AngleSharp pools objects like `StringBuilder`, which is great. However, access to these is locked, to make `Pool` thread safe.

An alternative might be to use `ThreadStatic<>` to have a `Pool` per thread. The only caveat is that pooled objects need to be accessed entirely within the same thread - they can't be handed off to another thread (which I don't see ever happening.)

Feel free to close this if the potential win is not worth the effort.

**Comment by FlorianRappl** [comment]
Yes you are right - the `ThreadStaticAttribute` could be used as well (there is no handover happening internally afaik). However, in the most critical paths the pool is only used very rarely (acquire as soon as possible and return as late as possible) reducing the overhead to a negligible constant.

The pool is also used for some things like serialization. Here I don't know how much "creating of a new pool once" (different thread) vs. "always synchronize access" (independent of thread) is. I assume a `ThreadStatic` could be beneficial. I will try to investigate it with some serialization trials.

At this point I am in favor.

**Comment by FlorianRappl** [comment]
I think I'll go for the `ThreadStaticAttribute` solution. The reasoning is simple: Currently every borrow / return action is more expensive than required. Considering that usually all actions are placed on the same thread there is not much performance to lose by potentially starting a new pool for _other_ threads.

I still need to come up with a good benchmark though.

**Comment by corliss** [comment]
Cool. However after a little [research](http://stackoverflow.com/a/4673642) it seems the overhead of lock is just 50ns when there is no contention. So I think either way may end up with exactly the same performance (and would be interested to find out if it is anything else.)

**Comment by FlorianRappl** [comment]
I tested it out and I am convinced that the current way is alright. I'll leave it as it is.

**Comment by corliss** [comment]
I agree. Also I just realized that [ThreadStatic] might itself have some significant overhead - after all how does .net associate each thread with a variable? Sure enough, this [post](http://tips.x-tensive.com/2008/10/cost-of-threadstatic-attribute.html) describes the 14x overhead caused by HashTable.

.net teaches something new every day :)

**Comment by douglasg14b** [comment]
@FlorianRappl 

I'm trying to determine where multithreaded my application is bottlenecking, and am finding the largest chunk of resource contention to be stemming from AngleSharp. Specifically on `HtmlParser.Parse()`. It's enough contention there that trying to split a parsing workload between multiple tasks results in worse performance overall.

Is that related to this? Or something else?

**Comment by FlorianRappl** [comment]
Is there a specific reason for using `Parse` instead of `ParseAsync`? Independent of that question yes, it may indeed be related, but I cannot answer for sure without knowing more about the application / workloads.

**Comment by douglasg14b** [comment]
@FlorianRappl Gotcha.

I'm used to `XAsync` methods actually just being CPU bound operations that abuse the `async` modifier, with no actual benefit, especially when you're already performing the operation on it's own thread and it doesn't matter if it gets blocked. I assumed this was no different since parsing should be a synchronous CPU bound operation on a single thread? (No offense, genuinely curious)

I can make an issue to describe what I'm running into with more details if needed. I don't entirely know how AngleSharp parses the HTML (Does it run synchronously from start to finish, use parallel processing, tasks...etc) so It's hard to debug.

**Comment by FlorianRappl** [comment]
While this is true for the "pure" parsing part there are several other parts that actually benefit a lot from the Async. It mostly shines with external resources / script exection and other things, but it becomes already useful regarding events that need to be emitted according to the spec. As I don't know your config or use case it is hard to determine - hence the question.

**Comment by douglasg14b** [comment]
@FlorianRappl 

Here is an overview of how I'm using this: I'm scraping forum threads from a website.

### Use 

I have several Tasks retrieving forum threads, and other Tasks parsing/manipulating, and a separate Task for writing to the DB. A forum thread is downloaded, I then queue it up and a processing task dequeues it and passes it into the HtmlParser. I then grab the values of elements I want (thread title, post bodies, timestamps, usernames, # of pages...etc), convert it into an appropriate Entity for Entity Framework then ensure all references to the original objects are cleaned up so they can be GCd.

### The Problem 

The problem is on the parsing side of it. One Task provides the highest performance for this, each tasks multiplies the time it takes to parse by the number of tasks I have. (1 Task = 20ms, 2 task = 42ms, 3 task = 65ms each). The parsing also annihilates the performance of any other task (like the downloading tasks, which eat almost no CPU), without increasing the system CPU utilization by much.

I decided to isolate just the HtmlParser. I queued up a bunch of forum threads, and ran them through no more than `parser.parse(source)`. I saw the same performance drop when trying to parse threads in parallel.

### Profiling:

Performance profiling time. I ran the concurrency profiler, and found an insane amount of contention right at `parser.parse(source)`, and a little bit at some points where I'm using CSS selectors to grab elements. I downloaded the [concurrency visualizer ](https://docs.microsoft.com/en-us/visualstudio/profiling/concurrency-visualizer) and found that during parsing the worker threads are ~79% blocked by synchrononization, and when looking at the Core view, each thread was performing hundreds of context switches, many to different cores during this time.

The short of it is I'm not gaining any performance benefits from trying to parse parallel HTML pages, and I'm not sure why. There may definitely be flaws in my methodology, but it's the best I have thus far.

Threads View:

![image](https://user-images.githubusercontent.com/1400380/28928874-4cc0ee52-7823-11e7-9950-174f434c1529.png)

Core View:

![image](https://user-images.githubusercontent.com/1400380/28928886-53d0580e-7823-11e7-88ae-695eeb4daad2.png)
